hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -374,93 +376,87 @@ private void updateKeys(
+                    } else {
+                        // With "full" verification mode, our required key
+                        // structure is a 1/2 threshold with components:
+                        //   - Admin key
+                        //   - A 2/2 list including the role key and its replacement key
+                        final var key = tokenKey.getFromUpdate(op);
+                        requireAdminOrRole(context, originalToken, tokenKey, key);
+                    }
                 }
-
-                context.requireKey(requiredKey);
-            } else if (!isExpiryOnlyUpdateOp(op)) {
-                // if we update any other field, we need the current admin key to sign
-                // note: for expiry only op admin key is not required
-                context.requireKey(originalToken.adminKey());
             }
-        } else {
-            lowPriorityKeys = getAllRequiredLowPrioritySigners(originalToken, op);
-            final Key lowPriorityKeyList = generateKeyListKey(lowPriorityKeys);
-            context.requireKey(lowPriorityKeyList);
         }
     }
 
-    private List<Key> getAllRequiredLowPrioritySigners(
-            @NonNull final Token originalToken, @NonNull final TokenUpdateTransactionBody op) {
-        List<Key> lowPriorityKeys = new ArrayList<>();
+    private void requireAdminOrRole(
+            @NonNull final PreHandleContext context, @NonNull final Token token, @NonNull final TokenKey roleKey)
+            throws PreCheckException {
+        requireAdminOrRole(context, token, roleKey, null);
+    }
 
-        for (final var nonAdminKey : NON_ADMIN_TOKEN_KEYS) {
-            if (nonAdminKey.isPresentInUpdate(op)) {
-                final Key presentKey = nonAdminKey.getFromToken(originalToken);
-                if (presentKey != null) {
-                    lowPriorityKeys.add(presentKey);
-                }
-            }
-        }
-        // metadata field can be updated also with only metadata key signature
-        if (op.hasMetadata()) {
-            final Key metadataKey = originalToken.metadataKey();
-            if (metadataKey != null) {
-                lowPriorityKeys.add(metadataKey);
+    private void requireAdminOrRole(
+            @NonNull final PreHandleContext context,
+            @NonNull final Token token,
+            @NonNull final TokenKey roleKey,
+            @Nullable final Key replacementKey)
+            throws PreCheckException {
+        final var maybeRoleKey = roleKey.getFromToken(token);
+        mustExist(maybeRoleKey, roleKey.tokenHasNoKeyStatus());
+        if (token.hasAdminKey()) {
+            context.requireKey(oneOf(
+                    replacementKey == null ? maybeRoleKey : allOf(maybeRoleKey, replacementKey),
+                    token.adminKeyOrThrow()));
+        } else {
+            context.requireKey(maybeRoleKey);
+            if (replacementKey != null) {
+                context.requireKey(replacementKey);
```

Comment:
Whenever a low priority key wants to update itself to other key. In this case, the new key is **not** required to sign the transaction
i.e: `kycKey` wants to update itself to `newKycKey` -> we require only `kycKey` signature (part of the HIP requirements)

Reply:
I have updated the HIP to clarify that without the admin signature, the new low-privilege key also needs to sign.

(This is the only meaningful interpretation of `TokenKeyValidation.NONE`, I think.)

Reply:
I see, this changes the initial requirements quite a bit

Reply:
True. :/